walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,105 @@
-
+import java.util.Collections;
+/* Implementation requires that Collection of Events are all sorted in order of meeting start time */
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> avaliableTimes = new ArrayList<>();
+
+    /* If the requested time span is longer than a whole day, return no avaliable meeting times */
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return avaliableTimes;
+    }
+
+    int blockStart = TimeRange.START_OF_DAY;
+    int blockStartOpt = TimeRange.START_OF_DAY;
+    Collection<String> requestAtten = request.getAttendees();
+    Collection<String> optionalAtten = request.getOptionalAttendees();
+
+    /* If there are only optional attendees requested, default set all optional attendeeds to be included */
+    boolean optionalIncluded = request.getAttendees().isEmpty();
+    
+    for (Event event : events) { 
+      /* Check if there are attendees in the event that are also requested attendees */
+      Collection<String> eventAtten = event.getAttendees();
+      boolean timeConflictFound = !Collections.disjoint(eventAtten, requestAtten);
+      boolean timeConflictFoundOpt = !Collections.disjoint(eventAtten, optionalAtten);
+      int eventStart = event.getWhen().start();
+      int eventEnd = event.getWhen().end();
+
+      /* If no optional attendee conflict and event length long enough, include optional attendee */
+      if (!timeConflictFoundOpt && eventStart - blockStartOpt >= request.getDuration()) {
+        optionalIncluded = true;
+      }
+      else if (timeConflictFoundOpt) {
+        blockStartOpt = eventEnd;
+      }
+
+      if (optionalIncluded) {
+        timeConflictFound =  timeConflictFound || timeConflictFoundOpt;
+      }
+
+      if (timeConflictFound) {
+        /* If time conflict found, add avaliable time block up until the start of the event */
+        avaliableTimes = checkConflict(avaliableTimes, request, blockStart, eventStart);
+
+        /* If avaliable start time is before the end of the conflict time block, 
+           set new avaliable time to after the conflict event is finished. */
+        if (blockStart < eventEnd) {
+          blockStart = eventEnd;
+        }
+      }
+    }
+    /* After all events considered, add the entire rest of the day as also avaliable */
+    avaliableTimes = checkEndOfDay(avaliableTimes, blockStart);
+
+    /* Is there are no avaliable times returned, but there are non 
+       optional requested attendees, check again for non optional 
+       attendee avaliability */
+    if (avaliableTimes.isEmpty() && !request.getAttendees().isEmpty()) {
+      avaliableTimes = queryAvailOnly(events, request);
+    }
+    return avaliableTimes;
+  }
+
+  private Collection<TimeRange> queryAvailOnly(Collection<Event> events, MeetingRequest request) {
```

Comment:
With this set up of first checking for times with optional attendees and then for non-optional attendees, you end up having to loop through all of the events twice. I won't block you on this, but it might be worthwhile to explore either in this PR or in a follow-up of consolidating the logic into one loop.